src/interpreter/balances.rs

```diff
@@ -0,0 +1,191 @@
+        Self { value, offset }
+    }
+
+    pub const fn offset(&self) -> usize {
+        self.offset
+    }
+
+    pub const fn value(&self) -> Word {
+        self.value
+    }
+
+    pub fn checked_add(&mut self, value: Word) -> Option<&mut Self> {
+        self.value.checked_add(value).map(|v| self.value = v).map(|_| self)
+    }
+
+    pub fn checked_sub(&mut self, value: Word) -> Option<&mut Self> {
+        self.value.checked_sub(value).map(|v| self.value = v).map(|_| self)
+    }
+}
+
+/// Structure to encapsulate asset balances for VM runtime
+#[derive(Debug, Default, Clone)]
+pub struct RuntimeBalances {
+    state: HashMap<AssetId, Balance>,
+}
+
+impl From<&CheckedTransaction> for RuntimeBalances {
+    fn from(tx: &CheckedTransaction) -> Self {
+        let iter = tx.free_balances().map(|(asset, value)| (*asset, *value));
+
+        Self::try_from_iter(iter).expect(
+r#"This is a bug!
+
+A checked transaction shouldn't produce a malformed initial free balances set.
+
+Please, file a report mentioning an incorrect transaction validation implementation that allowed a type-safe checked transaction to be created from a malformed inputs set."#)
+    }
+}
+
+impl RuntimeBalances {
+    /// Attempt to create a set of runtime balances from an iterator of pairs.
+    ///
+    /// This will fail if, and only if, the provided asset/balance pair isn't consistent or a
+    /// balance overflows.
+    pub fn try_from_iter<T>(iter: T) -> Result<Self, ValidationError>
+    where
+        T: IntoIterator<Item = (AssetId, Word)>,
+    {
+        iter.into_iter()
+            .sorted_by_key(|k| k.0)
+            .enumerate()
+            .try_fold(HashMap::new(), |mut state, (i, (asset, balance))| {
+                let offset = VM_MEMORY_BALANCES_OFFSET + i * (AssetId::LEN + WORD_SIZE);
+
+                state
+                    .entry(asset)
+                    .or_insert(Balance::new(0, offset))
+                    .checked_add(balance)
+                    .ok_or(ValidationError::ArithmeticOverflow)?;
+
+                Ok(state)
+            })
+            .map(|state| Self { state })
+    }
+
+    /// Fetch the balance of a given Id, if set.
+    pub fn balance(&self, asset: &AssetId) -> Option<Word> {
+        self.state.get(asset).map(Balance::value)
+    }
+
+    fn _set_memory_balance(balance: &Balance, memory: &mut [u8]) -> Word {
+        let value = balance.value();
+        let offset = balance.offset();
+
+        let offset = offset + AssetId::LEN;
+        let memory = &mut memory[offset..offset + WORD_SIZE];
+
+        memory.copy_from_slice(&value.to_be_bytes());
+
+        value
+    }
+
+    /// Attempt to add the balance of an asset, updating the VM memory in the appropriate
+    /// offset
+    pub fn checked_balance_add(&mut self, memory: &mut [u8], asset: &AssetId, value: Word) -> Option<Word> {
+        self.state
+            .get_mut(asset)
+            .and_then(|b| b.checked_sub(value))
```

Comment:
How did this not get caught in tests?

```suggestion
            .and_then(|b| b.checked_add(value))
```

Reply:
The reason this nit wasn't caught in tests is that a balance add never happen during VM runtime. We initialize the balances from the available free assets, and only decrease them.

But you're right, we should have this covered. I fixed it here and added unit tests to this new struct:

https://github.com/FuelLabs/fuel-vm/pull/173/commits/902e894408a347987f157371eea947322ecb6352